cheatsheets/Password_Storage_Cheat_Sheet.md

```diff
@@ -1,567 +1,153 @@
 
-2. Load and use new protection scheme
-    1. Load a new, stronger credential protection scheme (See next section on: Upgrading your existing password hashing solution)
-    2. Include version information stored with form
-    3. Set 'tainted'/'compromised' bit until user resets credentials
-    4. Rotate any keys and/or adjust protection function parameters such as work factor or salt
-    5. Increment scheme version number
+In some cases, it may be possible to increase the work factor of the hashes without the original password, although this is not supported by common hashing algorithms such as Bcrypt and PBKDF2.
 
-3. When user logs in:
-    1. Validate credentials based on stored version (old or new); if older compromised version is still active for user, demand 2nd factor or secret answers until the new method is implemented or activated for that user
-    2. Prompt user for credential change, apologize, & conduct out-of-band confirmation
-    3. Convert stored credentials to new scheme as user successfully log in
+## Maximum Password Lengths
 
-## Upgrading your existing password hashing solution
+Some hashing algorithms such as Bcrypt have a maximum length for the input, which can [vary between different implementations](https://security.stackexchange.com/questions/39849/does-bcrypt-have-a-maximum-password-length) of the algorithm. Where Bcrypt is used, a maximum length of 50 characters should be enforced on the input to cover all known implementations.
 
-The above guidance describes how to do password hashing correctly/safely. However, it is very likely you'll be in a situation where you have an existing solution you want to upgrade. This [article](https://veggiespam.com/painless-password-hash-upgrades/) provides some good guidance on how to accomplish an upgrade in place without adversely affecting existing user accounts and future proofing your upgrade so you can seamlessly upgrade again (which you eventually will need to do).
+Additionally, due to how computationally expensive modern hashing functions are, if a user can supply very long passwords then there is a potential denial of service vulnerability, such as the one published in [Django](https://www.djangoproject.com/weblog/2013/sep/15/security/) in 2013.
 
-# Argon2 usage proposal in Java
+In order to protect against both of these issues, a maximum password length should be enforced. This should be 50 characters for Bcrypt, and at least 64 characters for other algorithms. There are two main ways that this can be done:
 
-The objective is to propose a example of secure usage/integration of the Argon2 algorithm in Java application to protect password when stored.
+- Restrict the maximum length in a password policy.
```

Comment:
I replied to this a week ago in Slack, but never got a response, so copying my reply here so it doesn't get lost in Slack's history.

> I don't really see a much difference between pre-hashing the password and enforcing a maximum password length on the client side in terms of the overall security provided.
> 
> If you pre-hash the password, you should encode it before passing it to the bcrypt function (as bcrypt may act strangely if there are null bytes in the input). You can either use SHA-1 (40 hex characters output), or SHA-256 (43 base64 characters) - to go above that you would need to do some kind of custom encoding to stay safely under the 50 characters.
> 
> The hash function is effectively reducing the keyspace for the bcrypt input to 2^160 or 2^256 (depending on the hash). 50 mixed-alpha-numeric-symbol characters has a keyspace of 95^50 (approximately 2^328) - which is actually a much bigger keyspace that you get from either of the hash functions.
> 
> To my mind, by allowing arbitrary length passwords and pre-hashing, you're not really providing any more security, and you're making the authentication code more complex than it needs to be for no real benefit.
> 
> Are there any circumstances where pre-hashing is better than just limiting the maximum password length to 50 characters? The only reason that I can think of is that it might "look better" not to have the arbitrary limit - but if you're that concerned about security then you should be doing something practical to improve it (like using Argon2, which doesn't have the length limitation), rather than trying to hide it with pre-hashing.

If anyone has any technical arguments for/against this approach I'd be interested to here them either here or on Slack.